sound/soc/sof/ipc4-topology.c

```diff
@@ -204,6 +213,22 @@ static int sof_ipc4_setup_process_init_byte_kcontrol(struct snd_sof_widget *swid
 		process->cdata = control_data;
 		break;
 	}
+
+	module_lib_id = sof_ipc4_get_module_lib_id(swidget->module_info);
+	/* module_lib_id > 0 if module is not in basefw and loaded from external library */
+	if (module_lib_id && process->cdata) {
```

Comment:
it is already done in the bove loop @ujfalusi

Reply:
> if (process->cdata && process->cdata->blob_type == SOF_IPC4_MOD_INIT_INSTANCE)

This can not guarantee we have the base_cfg_ext struct in the data section, because some basefw modules also has the cdata for module init and supported with process ops. mic_sel module is one of such.
```
struct selector_comp_data {
    struct ipc4_audio_format output_format;
    struct ipc4_selector_coeffs_config coeffs_config;
}
```
coeffs_config is provided from topology as byte kontrol.
https://github.com/thesofproject/sof/blob/82220c823c1a3783e1a21eb5e02fb9058a64ed94/src/include/sof/audio/selector.h#L71

Reply:
@aiChaoSONG, I see, but with the mention of mic_sel this all makes even less sense and it is mixing up several things.
For mic_sel the blob (from tplg) is the coefficient configuration, the audio_fmt is _not_ part of the blob.
The blob is not parsed by the kernel, but it will create the audio_fmt+blob for the firmware internally.

With this new blob type you are introducing, the audio fmt is also part of the blob, which forces the kernel to parse a blob (which by definition is an opaque binary thing).

Furthermore: I still don't see how the base_module_cfg_ext is mandatory and only possible with modules not in the basefw...

What would happen with the mic_sel module if it is moved out to a separate library since we don't need it to be taking up code space on setups where it is not used? It's mod_init_instance does not need module_base_cfg_ext at all, but it needs audio_fmt+ coef and the topology itself contains only the coef as binary. The topology has no idea if the module with the given UUID is in the basefw or not. If it is not in basefw, but the blob does not include firmware message portion (module_base_cfg_ext) then you can not parse the blob based on module_lib_id, the blob still does not contain the base_cfg_ext.

Is it an oversimplification to say that some module is using base_cfg_ext in mod_init_instance to save on extra IPC to configure extra pins and provide single message way to pass the configuration compared to 'old school' modules which needed extra config message to set the configuration (or modules which can be reconfigured while running - new coef for example).

In other words: there are modules which uses mod_init_instance + some sort of config message (MOD_CONFIG/LARGE_CONFIG) and some with a single mod_init_instance with base_cfg_ext ?
The ones with mod_init_instance can not be reconfigured runtime, they need to be stopped and restarted to be able to use the new configuration?

How would a module which was developed as external and used the base_cfg_ext would behave and work if it is moved inside of basefw? How it would get the configuration it needs for operating (which is expected via mod_init_instance message)?

How I see is that the base_cfg_ext and the fact that the module is not in basefw as accidental and not intentional, I don't think that modules not in the basefw must only support mod_init_instance with base_cfg_ext and modules in basefw cannot support this combination of initialization.

Yes, this is the case at the moment, but the rule that we are trying to introduce is wrong, it is per module parameter and not location (in basefw or outside of basefw)

Reply:
@aiChaoSONG, it might be 'small' modification to generate a blob with the base_cfg_ext, but then what? How do you know which blob to send and how do you make sure that the correct blob is in the tplg (and user space) for the given firmware configuration for the module (it can be built in or external)?
Have this base_cfg_ext always in the blobs and strip it off in kernel if not needed? But if the module needs this info?

Reply:
After syncing with the firmware side, the assumption that the base config extension structure is only used with external module is incorrect, it could be both used by built-in modules or external modules. 

I made the assumption previsouly, because SOF loadable feature is not yet ready,  and only the iadk modules used the library manager code in sof, and we also have a repo for iadk modules: https://github.com/thesofproject/converged-sof-modules. I thought we will support loadable feature with IADK infra.

Good lesson for me on cross-team communication. I think we need a token to indicate which blob format is used now, let me bring the token in the initial version of this PR back.